### PR TITLE
[WIP] Smoke test with docker

### PR DESCRIPTION
**Pull request by camallen**
Get started using docker images via docker and compose. this cleans up the dockerfile to rely on the `bootstrap.sh` script and allows dev via the 

It's not optimal in that we copy the src tree into the image before setting up the deps, ideally we'd breakup the bootstrap.sh script to avoid this bulk copy (happy to help on this but might need some direction). 

# Review Checklist

## General

- [ ] Are the tests passing locally and on Travis?
- [ ] Is the documentation up to date?
- [ ] Is the changelog updated?

## Apps

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `rm -rf node_modules/ && npm install` and app works as expected?

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,13 +1,12 @@
...
+ADD . /usr/src
+
 RUN chown -R node:node .
 
 USER node
 
-RUN npm install
-RUN ./node_modules/.bin/lerna link
-RUN ./node_modules/.bin/lerna bootstrap --no-ci --loglevel verbose
+RUN npm config set unsafe-perm true
```
Comment (camallen) [inline Dockerfile]: i wasn't' super keen to use this config options but couldn't get it building without...thoughts appreciated.

Comment (eatyourgreens) [reply Dockerfile]: We use the same config to build the JS for zooniverse.org.
https://github.com/zooniverse/Panoptes-Front-End/blob/fb066b66f67f02c759b42b4ff95dba0564c56c39/Dockerfile#L13

Set it to false if npm is run as root, but that sounds less safe to me.
https://docs.npmjs.com/misc/config#unsafe-perm

**Comment by eatyourgreens** [comment]
Can we build and start a local server, serving app-project, with `docker-compose up`? That'd be convenient compared to running the app locally at the moment.

**Comment by eatyourgreens** [comment]
The bootstrap script symlinks all the local packages so that npm can install them. Once they're published to a repo (I assume npm) then you shouldn't have to mount the whole packages tree just to run the project app.

**Comment by camallen** [comment]
> Can we build and start a local server, serving app-project, with `docker-compose up`? That'd be convenient compared to running the app locally at the moment.

Yep can do - I couldn't see any commands to do this sort of thing in the readme. What are the commands to run the local dev server? Does it do hot code reloading?

**Comment by camallen** [comment]
> The bootstrap script symlinks all the local packages so that npm can install them. Once they're published to a repo (I assume npm) then you shouldn't have to mount the whole packages tree just to run the project app.

I'm not sure what this means, is this for production deployment or  development? In dev i assume we will have to mount the packages to get the build working? The whole system was a little tricky to build tbh. 

I can split the dockerfile to a dockerfile.dev for local development work vs published packages for production..

**Comment by eatyourgreens** [comment]
```sh
cd packages/app-project
npm run dev
```
after running the bootstrap script, should start a local server on port 3000.

None of the packages are published (@rogerhutchings or @srallen can correct me here) so the bootstrap script symlinks them to avoid 'module not found' errors when `npm install` runs and tries to install dependencies like `@zooniverse/react-components`.

I can't remember if the bootstrap script builds any packages that need to be built before they can be imported.

**Comment by srallen** [comment]
The bootstrap script also builds the packages. Whether or not docker should run this script depends on what the goal is. If it is development, then I think it should. Reason being is that you're updating the grommet theme package or the react-components and you need to then test in the classifier or app-project. These changes should all be done locally and then built and symlinked for testing.

**Comment by camallen** [comment]
I want to provide a way for the system to build in production mode for deployment as well. As the packages aren't published to npm yet we still need to build them locally.

I'm going to split out the dockerfiles to prod and dev, ideally the production version will just `npm install` to get what it needs to run the ci tests and deploy to swarm. 

Development mode will build locally (ideally knowing what has changed / it needs to build) and inject that into a running dev environment.

**Comment by rogerhutchings** [comment]
This is pretty out of date by now, but I'm planning to do this next, so I'm going to close the PR but leave the branch.

**Pull request closed without merging**